src/azure-cli/azure/cli/command_modules/appconfig/_kv_helpers.py

```diff
@@ -879,23 +852,36 @@ def __export_keyvalues(fetched_items, format_, separator, prefix=None):
                 continue
 
             key_segments = key.split(separator)
-            if key_segments[0].isdigit():
-                __export_keyvalue(key_segments,
-                                  kv.value, exported_list, key)
-
-            else:
-                __export_keyvalue(key_segments,
-                                  kv.value, exported_dict, key)
-
-        if exported_dict and exported_list:
-            logger.error("Can not export to a valid file! Some keys have been dropped. %s", json.dumps(
-                exported_dict, indent=2, ensure_ascii=False))
+            __export_keyvalue(key_segments, kv.value, exported_dict)
 
-        return __compact_key_values(exported_dict if not exported_list else exported_list)
+        return __try_convert_to_arrays(exported_dict)
     except Exception as exception:
         raise CLIError("Fail to export key-values." + str(exception))
 
 
+def __try_convert_to_arrays(constructed_data, is_root=True):
+    if not (isinstance(constructed_data, dict) and len(constructed_data) > 0):
+        return constructed_data
+
+    # Object cannot be an array if not all keys are numeric
+    if False not in (key.isdigit() for key in constructed_data):
+        is_array = True
+        sorted_data_keys = sorted(int(key) for key in constructed_data)
+
+        # If all keys are digits and in order starting from 0, we convert the dictionary to an array
+        # with indices corresponding to the keys.
+        # We do not try to convert key-values at the root of the object to an array even if they meet this criterion.
+        for index, key in enumerate(sorted_data_keys):
+            if is_root or index != key:
```

Comment:
>index != key [](http://example.com/codeflow?start=26&length=12)

We also support converting something like this to a valid array:
```
arr:0='abc'
arr:6='pqr'
arr:7='xyz'
```

This would still flatten to `arr = ["abc","pqr","xyz"]`

Reply:
We ensure that the indices are in continuous order from 0 to n else we revert to the default export as key-values. This is because a value could be easily inserted into the array by a user if they add, e.g 
arr:4="efg" separately after importing these values.  This changes the indices of values for keys "arr:6" and "arr:7" in the exported array.  However, the indices are guaranteed to stay the same if the values are continuous and one could only append to the end of the array by adding a key continuous with the current order.

Reply:
That's a valid scenario that we have always supported. If someone adds `arr:4="efg"` after importing key-values, the final result after exporting should be `arr = ["abc","efg","pqr","xyz"]`

Consider this scenario - someone imports an array in their config store and then they have to delete some values from the middle of the array. Do we expect them to manually reorder all their key-values just because they deleted one value in the middle? In this scenario they will end up with the example I described above, and we shouldn't convert their array to key-values just because the indices are not contiguous.

Reply:
That is a valid argument, however, int this same scenario, that means we will be changing key names assuming the exported file is imported to a different key-value store. Or there could be duplicates if we import back to the same key-value store for any reason since now a value that was initially `arr:4` will now be `arr:2`.  This will not be easily caught in a user's code if they are accessing a value in a given index. e.g

A user is accessing `arr:2` in `arr = ["abc","efg","pqr","xyz"]` from their code. When we delete this value, the value of key `arr:2` and all other indices after that are going to change, resulting in unexpected behaviour which will be difficult to trace. But if we revert to key-values, the code will just break since the configuration will not have been found.

I think exporting as key-values is the default mode which can be consistently exported/imported so we revert to that to preserve them. But we could discuss this further.